### PR TITLE
planner: refactor handle columns

### PR DESCRIPTION
**Pull request by coocood**
### What problem does this PR solve?

Make the clustered index code easier to maintain.

### What is changed and how it works?

Define an interface `HandleCols` that abstracts the handle column for common handle and int handle.

### Check List 

Tests 

- refactor covered by existing tests.

Side effects

### Release note 

- No release note

### CODE REVIEW

File: planner/core/logical_plan_builder.go [hunk 1]
```diff
@@ -2965,14 +2964,14 @@ func (b *PlanBuilder) buildMemTable(_ context.Context, dbName model.CIStr, table
...
 
-	if handleCol != nil {
-		handleMap := make(map[int64][]*expression.Column)
-		handleMap[tableInfo.ID] = []*expression.Column{handleCol}
+	if handleCols != nil {
+		handleMap := make(map[int64][]HandleCols)
+		handleMap[tableInfo.ID] = []HandleCols{handleCols}
 		b.handleHelper.pushMap(handleMap)
 	} else {
 		b.handleHelper.pushMap(nil)
```
Comment (lysu) [inline planner/core/logical_plan_builder.go:2977]: do we need check `tableInfo.isCommaHandle ` and set `handleCols` just like `buildDataSource`~?

File: planner/core/logical_plan_builder.go [hunk 2]
```diff
@@ -3736,33 +3735,40 @@ func (b *PlanBuilder) buildDelete(ctx context.Context, delete *ast.DeleteStmt) (
...
 			delete(tblID2Handle, id)
 			continue
 		}
 		for i := len(cols) - 1; i >= 0; i-- {
-			if !p.matchingDeletingTable(names, outputNames[cols[i].Index]) {
+			hCols := cols[i]
+			var hasMatch bool
+			for j := 0; j < hCols.NumCols(); j++ {
+				if p.matchingDeletingTable(names, outputNames[hCols.GetCol(j).Index]) {
+					hasMatch = true
```
Comment (lysu) [inline planner/core/logical_plan_builder.go:3768]: maybe we can break after this line

File: planner/core/logical_plan_builder.go (hunk 1 above)
Comment (coocood) [reply planner/core/logical_plan_builder.go:2977]: mem table would never be common handle.

File: planner/core/rule_column_pruning.go [hunk 3]
```diff
@@ -221,17 +223,13 @@ func (ds *DataSource) PruneColumns(parentUsedCols []*expression.Column) error {
...
-	)
-	if ds.handleCol != nil {
-		handleCol = ds.handleCol
-		handleColInfo = ds.Columns[ds.schema.ColumnIndex(handleCol)]
-	}
 	originSchemaColumns := ds.schema.Columns
 	originColumns := ds.Columns
 	for i := len(used) - 1; i >= 0; i-- {
+		if ds.tableInfo.IsCommonHandle && mysql.HasPriKeyFlag(ds.schema.Columns[i].RetType.Flag) {
+			// Do not prune common handle column.
```
Comment (lzmhhh123) [inline planner/core/rule_column_pruning.go:230]: Why we don't prune common columns in schema?

File: planner/core/rule_max_min_eliminate.go [hunk 4]
```diff
@@ -61,11 +61,11 @@ func (a *maxMinEliminator) checkColCanUseIndex(plan LogicalPlan, col *expression
 	case *DataSource:
 		// Check whether there is an AccessPath can use index for col.
 		for _, path := range p.possibleAccessPaths {
-			if path.IsTablePath() {
+			if path.IsIntHandlePath {
```
Comment (lzmhhh123) [inline planner/core/rule_max_min_eliminate.go:64]: Add a `TODO` here to support `commonHandlePath`?

File: planner/core/testdata/integration_suite_out.json [hunk 5]
```diff
@@ -848,8 +848,9 @@
       {
         "SQL": "select t1.b, t1.c from t1 where t1.c = 2.2",
         "Plan": [
-          "IndexReader_6 1.00 root  index:IndexRangeScan_5",
-          "└─IndexRangeScan_5 1.00 cop[tikv] table:t1, index:c(c) range:[2.2000000000,2.2000000000], keep order:false"
+          "Projection_4 1.00 root  test.t1.b, test.t1.c",
```
Comment (lzmhhh123) [inline planner/core/testdata/integration_suite_out.json:851]: It seems the projection is not necessary.

File: planner/core/find_best_task.go [hunk 6]
```diff
@@ -870,11 +870,11 @@ func (ds *DataSource) isCoveringIndex(columns, indexColumns []*expression.Column
 
 // If there is a table reader which needs to keep order, we should append a pk to table scan.
 func (ts *PhysicalTableScan) appendExtraHandleCol(ds *DataSource) (*expression.Column, bool) {
-	handleCol := ds.handleCol
-	if handleCol != nil {
-		return handleCol, false
+	handleCols := ds.handleCols
+	if handleCols != nil {
```
Comment (lzmhhh123) [inline planner/core/find_best_task.go:874]: May this check be `false`? I think it's always `not nil`.

Comment (coocood) [reply planner/core/find_best_task.go:874]: handleCols may be set to nil in `DataSource.PruneColumns`.

File: planner/core/rule_column_pruning.go (hunk 3 above)
Comment (coocood) [reply planner/core/rule_column_pruning.go:230]: For simplicity.

**Pull request merged**